File: day4/day4.ts
import { getExampleInput, getPuzzleInput } from "../helper/inputs";

const example: string[] = getExampleInput(__dirname);
const input: string[] = getPuzzleInput(__dirname);

console.log(`Part 1 example: ${solve1(example)}`);
console.log(`Part 1 solution: ${solve1(input)}`);
console.log(`Part 2 example: ${solve2(example)}`);
console.log(`Part 2 solution: ${solve2(input)}`);

function solve1(lines: string[], sum = 0) {
  lines.forEach((line) => {
    let cCard: Card = parseCard(line);
    sum = sum + calculatePoints(cCard);
  });

  return sum;
}

function solve2(lines: string[], sum = 0) {
  let gameids = [];
  lines.forEach((line) => {
    let cCard: Card = parseCard(line);

    gameids.push(cCard.id);
    let count = 0;
    cCard.my.forEach((n) => {
      if (cCard.win.includes(n)) {
        count++;
      }
    });

    for (let i = count; i > 0; i--) {
      let occ = countOccurrences(gameids, cCard.id);
      for (let j = occ; occ > 0; occ--) {
        gameids.push(cCard.id + i);
      }
    }
  });

  return gameids.length;
}

type Card = {
  id: number;
  my: string[];
  win: string[];
};

function calculatePoints(cCard: Card) {
  let count = 0;
  cCard.my.forEach((n) => {
    if (cCard.win.includes(n)) {
      count++;
    }
  });

  return (count > 0) ? 2 ** (count - 1) : 0
}

function countOccurrences<Type>(arr: Type[], val: Type): number {
  return arr.reduce((a, v) => (v === val ? a + 1 : a), 0);
}

function parseCard(l: string) {
  let [cid, numbers] = l.split(":");
  let [mine, winning] = numbers.split("|");
  cid = cid.replace("Card ", "");

  let cCard: Card = {
    id: +cid,
    my: mine.split(" ").filter((i) => i),
    win: winning.split(" ").filter((i) => i),
  };
  return cCard;
}

File: day4/day4.jl
include(pwd() * "\\helper\\inputs.jl")
example = getExample(@__DIR__)
input = getInput(@__DIR__)

function part1(lines)
    cards = scoreCards(parseCards(lines))
    sum(map(c -> c[4], cards))
end

function part2(lines)
    cards = scoreCards(parseCards(lines))
    countedCards = Dict(first.(cards) .=> 1)
    for (k, (_, _, _, _, won)) in pairs(cards)
        c = countedCards[k]
        for i in k+1:k +won
            countedCards[i] += 1*c
        end
    end
    sum(values(countedCards))
end

function scoreCards(cards)
    for (k, (_, winning, having)) in pairs(cards)
        won = winning ∩ having
        score = 0
        if length(won) > 0
            score = 2^(length(won) - 1)
        end
        append!(cards[k], score)
        append!(cards[k], length(won))
    end
    return cards
end

function parseCards(lines)
    lines = replace.(lines, "Card " => "")
    result = []
    for line in lines
        (id, tail) = split(line, ":")
        id = parse(Int, strip(id))
        (winning, having) = map(x -> strip(x), split(tail, "|"))
        winning = map(c -> parse(Int, c), split(winning))
        having = map(c -> parse(Int, c), split(having))
        push!(result, [id, winning, having])
    end
    return result
end

println("Part 1 example : $(part1(example))")
println("Part 1 solution: $(part1(input))")
println("Part 2 example : $(part2(example))")
println("Part 2 solution: $(part2(input))")
